lib/MusicBrainz/Script/RemoveUnreferencedRows.pm

```diff
@@ -0,0 +1,112 @@
+package MusicBrainz::Script::RemoveUnreferencedRows;
+use Moose;
+
+use DBDefs;
+use MusicBrainz::Server::Context;
+use MusicBrainz::Server::Log qw( log_info );
+
+with 'MooseX::Runnable';
+with 'MooseX::Getopt';
+with 'MusicBrainz::Script::Role::Context';
+
+has dry_run => (
+    isa => 'Bool',
+    is => 'ro',
+    default => 0,
+    traits => [ 'Getopt' ],
+    cmd_flag => 'dry-run'
+);
+
+has summary => (
+    isa => 'Bool',
+    is => 'ro',
+    default => 1,
+);
+
+has verbose => (
+    isa => 'Bool',
+    is => 'ro',
+    default => 0,
+);
+
+sub run {
+    my ($self) = @_;
+
+    my ($count, $removed) = (0, 0);
+
+    log_info { 'Checking all unreferenced rows' };
+
+    my @unreferenced_rows = @{
+        $self->c->sql->select_list_of_lists(<<~'SQL');
+              SELECT table_name,
+                     array_agg(row_id) AS row_ids
+                FROM unreferenced_row_log
+               WHERE inserted < now() - '7 day'::interval
+            GROUP BY table_name
+            SQL
+    };
+
+    for my $table_group (@unreferenced_rows) {
+        my ($table_name, $row_ids) = @$table_group;
+
+        log_info { sprintf "Checking rows for $table_name" };
+
+        # We want to ensure the rows are still unreferenced
+        my $query = <<~"SQL";
+            SELECT id
+              FROM $table_name
+             WHERE ref_count = 0
+               AND id = any(?)
+            SQL
+
+        my $unused_ids = $self->c->sql->select_single_column_array(
+            $query,
+            $row_ids,
+        );
+
+        for my $id (@$unused_ids) {
+            ++$count;
+
+            if ($self->dry_run) {
+                log_info {
+                    sprintf "Need to remove id=%s from $table_name",
+                    $id,
+                } if $self->verbose;
+            }
+            else {
+                Sql::run_in_transaction(sub {
+                    if ($table_name eq 'artist_credit') {
+                        $self->sql->do(<<~'SQL', $id);
+                            DELETE FROM artist_credit_gid_redirect
+                                  WHERE new_id = ?
+                            SQL
+                    }
+                    $self->c->sql->do(<<~"SQL", $id);
+                        DELETE FROM $table_name
+                              WHERE id = ?
```

Comment:
Expanding on my previous comment, one way to handle the race condition would be to do `WHERE id = ? AND ref_count = 0` here, then check `$self->c->sql->row_count` to see if anything was deleted.  But we'll have to `die` manually if nothing was deleted so that the `artist_credit_gid_redirect` deletion gets rolled back.  It might make sense to perform the `unreferenced_row_log` deletions individually in this transaction too.

Reply:
I understand this approach and the catch_exceptions one are not needed in combination? Or do you think they both make sense?

Reply:
Now that I looked at the latest code I think this is needed.  But it'd actually simplify the code (you wouldn't need the `$unused_ids` query at all) and be free of race conditions.  Maybe just keep a separate list of IDs that should be deleted from the log table at the end.

Reply:
I tried to use `$self->c->sql->row_count` but it's not working, getting this: `Error while processing id 1 from artist_credit: Cannot delegate row_count to rows because the value of sth is not defined at inline delegation in Sql for sth->rows`. Pushed it so you can maybe take a look.